src/parquet/ReadParquet.cpp

```diff
@@ -1,5 +1,407 @@
+
+// specialization for storing 32-bit data in 64-bit arrays
+template<> struct ChplBuf<arrow::Int32Type>  { using ElemType = int64_t; };
+template<> struct ChplBuf<arrow::UInt32Type> { using ElemType = uint64_t; };
+template<> struct ChplBuf<arrow::FloatType>  { using ElemType = double; };
+
+// TODO The implementation seemingly reads 128-bit decimal values into
+// Chapel `real`s (or C `double`s). I don't see how this can work properly with
+// large enough data. For now, I think this keeps the implementation practically
+// same as the old, likely carrying the same bug into the new implementation,
+// unfortunately.
+template<> struct ChplBuf<parquet::FLBAType>  {using ElemType = double;};
+
+// this struct is used to bundle different types into one, based on a single
+// ArrowType
+template<typename ArrowType>
+struct TypeBundle {
+  using ReaderType = typename parquet::TypedColumnReader<ArrowType>;
+  using ChplType = typename ChplBuf<ArrowType>::ElemType;
+  using PqType = typename ArrowType::c_type;
+};
+
+template<>
+struct TypeBundle<arrow::Decimal128Type> {
+  using ReaderType = typename parquet::TypedColumnReader<parquet::FLBAType>;
+  using ChplType = typename ChplBuf<parquet::FLBAType>::ElemType;
+  using PqType = typename parquet::FixedLenByteArray;
+};
+} // end anonymous namespace
+
+// Engin: We could consider exposing this struct to Chapel. Note that in
+// ParquetMsg.chpl there is a somewhat similar type. However, this one is for a
+// single column read
+// Based on what we decide, the implementation of this struct can be mode more
+// C++-like or could stay more C-like.
+struct AkColReadOp {
+  void* chpl_arr;
+  int64_t *startIdx;
+  std::shared_ptr<parquet::ColumnReader> column_reader;
+  bool hasNonFloatNulls;
+  chplEnum_t nullMode;
+  int64_t row_idx;
+  int64_t numElems;
+  int64_t batchSize;
+  bool* where_null_chpl;
+  const parquet::ColumnDescriptor* col_info;
+
+  // Generic read implementation
+  template<typename ArrowType>
+  int64_t read() {
+    using Types = TypeBundle<ArrowType>;
+    using ChplType = typename Types::ChplType;
+    using ReaderType = typename Types::ReaderType;
+
+    auto chpl_ptr = (ChplType*)chpl_arr;
+    int64_t num_read = 0;
+    ReaderType* reader = static_cast<ReaderType*>(column_reader.get());
+    *startIdx -= reader->Skip(*startIdx);
+
+    // TODO find better variable names. values_read and num_read are confusing
+    int64_t values_read = 0;
+
+    // note that floats are the types that natively support nulls. This generic
+    // read() shouldn't do anything with floats
+    if (nullMode == noNulls || nullMode == onlyFloats) {
+      while (reader->HasNext() && row_idx < numElems) {
+        if((numElems - row_idx) < batchSize) // adjust batchSize if needed
+          batchSize = numElems - row_idx;
+        reader->ReadBatch(batchSize, nullptr, nullptr, &chpl_ptr[row_idx],
+                          &values_read);
+        row_idx+=values_read;
+        num_read+=values_read;
```

Comment:
nit, formatting

Reply:
I will postpone this to the next refactor. It is against my style preferences, too. But if I am going in that direction, I would want to change variable names first.